README.md
# GiveMeAShot

EN - This addon automatically takes screenshots of specific events.

DE - Dieses Addon macht automatisch Screenshots zu bestimmten Ereignissen.


- [Download from Curse](https://www.curseforge.com/wow/addons/give-me-a-shot)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
[Added, Changed, Deprecated, Removed, Fixed, Security]


## [1.0.0] - YYYY-MM-DD
### Added
- First release


core.lua
GiveMeAShot = LibStub("AceAddon-3.0"):NewAddon("GiveMeAShot", "AceConsole-3.0", "AceEvent-3.0", "AceHook-3.0")
local addonName = GiveMeAShot:GetName()
--
local L = LibStub("AceLocale-3.0"):GetLocale("GiveMeAShot")
local Config, Util
--
local LDB = LibStub("LibDataBroker-1.1"):NewDataObject(addonName, {
  type = "launcher",
  text = L["MinimapButtonTooltipTitle"],
  icon = "Interface\\Icons\\trade_engineering",
  OnTooltipShow = function(tooltip)
    tooltip:SetText(L["MinimapButtonTooltipTitle"])
    tooltip:AddLine(L["MinimapButtonTooltipInfo"], 1, 1, 1)
    tooltip:Show()
  end,
  OnClick = function(self, button)
    if button == "LeftButton" then
      GiveMeAShot:ShowHelpFrame()
    elseif button == "RightButton" then
      GiveMeAShot:ShowOptionsFrame()
    end
  end})
local MiniMapButton = LibStub("LibDBIcon-1.0")


--------------------------------------------------
-- Variable definitions
--------------------------------------------------
local _state = {}


--------------------------------------------------
-- General Functions
--------------------------------------------------


--------------------------------------------------
-- Interface Events & Functions
--------------------------------------------------


--------------------------------------------------
-- Functions
--------------------------------------------------
function GiveMeAShot:ShowHelpFrame()
  -- double call to open the correct interface options panel -> Blizzard needs to fix
  InterfaceOptionsFrame_OpenToCategory(GiveMeAShot.helpFrame)
  InterfaceOptionsFrame_OpenToCategory(GiveMeAShot.helpFrame)
end

function GiveMeAShot:ShowOptionsFrame()
  -- double call to open the correct interface options panel -> Blizzard needs to fix
  InterfaceOptionsFrame_OpenToCategory(GiveMeAShot.optionsFrame)
  InterfaceOptionsFrame_OpenToCategory(GiveMeAShot.optionsFrame)
end

function GiveMeAShot:ToggleMinimapButton()
  Config:SetProfile('minimapButton.hide', not Config:GetProfile('minimapButton.hide'))
  if Config:GetProfile('minimapButton.hide') then
    MiniMapButton:Hide("GiveMeAShot")
  else
    MiniMapButton:Show("GiveMeAShot")
  end
end


--------------------------------------------------
-- Register Slash Commands
--------------------------------------------------
SLASH_RELOADUI1 = "/rl";
SlashCmdList.RELOADUI = ReloadUI;

function GiveMeAShot:ChatCommands(msg)
  local msg, msgParam = strsplit(" ", msg, 2)
  
  -- if msg == "minimap" then
  --   GiveMeAShot:ToggleMinimapButton()
  -- else
    GiveMeAShot:ShowOptionsFrame()
  -- end
end


--------------------------------------------------
-- Main Events
--------------------------------------------------
function GiveMeAShot:OnInitialize()
  -- -- setup popup dialogs
  -- self:SetupPopupDialogs()
end

function GiveMeAShot:OnEnable()
  -- load modules
  Config = self:GetModule('Config')
  Util = self:GetModule('Util')

  -- setup options frame
  Config:SetupOptionsUI();
  -- setup help site
  Config:SetupHelpUI();

  -- register minimap button
  MiniMapButton:Register("GiveMeAShot", LDB, Profile('minimapButton'))

  -- register slash commands
  self:RegisterChatCommand("givemeashot", "ChatCommands")

  -- Setup current character values
  Config:SetChar('level', UnitLevel("player"))
  Config:SetChar('race', select(2, UnitRace("player")))
  Config:SetChar('class', UnitClass("player"))
  Config:SetChar('class2', select(2, UnitClass("player")))
  Config:SetChar('faction', UnitFactionGroup("player"))
end

locale/deDE.lua
local L = LibStub("AceLocale-3.0"):NewLocale("GiveMeAShot", "deDE")
if not L then return end

-- Put the language in this locale here
L["Language"] = "Deutsch"


--------------------------------------------------
-- General
--------------------------------------------------
L["Addon_Title"] = "Give me a shot!"
L["Addon_Title_Short"] = "GiveMeAShot!"

L["Yes"] = "Ja"
L["No"] = "Nein"

L["PerformReload"] = "Das Interface muss neu geladen werden, damit 'Give me a shot' weiterhin ordnungsgemäß funktioniert.\n\nMöchtest du das UI jetzt neu laden?"
L["NotReloaded"] = "Das Interface wurde nicht neu geladen. Deine Änderungen sind noch nicht gültig."


--------------------------------------------------
-- UI elements
--------------------------------------------------
L["MinimapButtonTooltipTitle"] = "GiveMeAShot"
L["MinimapButtonTooltipInfo"] = "\124cffFF4500Linksklick:\124r Öffne die Hilfeseite.\n\124cffFF4500Rechtsklick:\124r Öffne das Optionsmenü."

L["MinimapButton"] = "Minikarten Schaltfläche"
L["MinimapButton_Desc"] = "Zeigt oder versteckt die Schaltfläche an der Minikarte."

L["DebugMode"] = "Debug Mode"
L["DebugMode_Desc"] = "Aktiviert oder deaktiviert den Debugmodus. Für normale Benutzer ist das nicht notwendig oder empfehlenswert."
L["DebugMode_Activated"] = "<Debug Mode> aktiviert."
L["DebugMode_Deactivated"] = "<Debug Mode> deaktiviert."

L["Profiles"] = "Profile"


--------------------------------------------------
-- Help
--------------------------------------------------
L["Help"] = "Hilfe"

L["Help_Opener"] = "Hier sollte ein Hilfetext stehen."


--------------------------------------------------
-- Error messages
--------------------------------------------------


locale/enUS.lua
local L = LibStub("AceLocale-3.0"):NewLocale("GiveMeAShot", "enUS", true)

-- Put the language in this locale here
L["Language"] = "English"


--------------------------------------------------
-- General
--------------------------------------------------
L["Addon_Title"] = "Give me a shot!"
L["Addon_Title_Short"] = "GiveMeAShot!"

L["Yes"] = "Yes"
L["No"] = "No"

L["PerformReload"] = "The interface needs to be reloaded for 'Give me a shot' to continue to work properly.\n\nDo you want to reload the UI now?"
L["NotReloaded"] = "The interface has not been reloaded. Your changes are not valid yet."


--------------------------------------------------
-- UI elements
--------------------------------------------------
L["MinimapButtonTooltipTitle"] = "GiveMeAShot"
L["MinimapButtonTooltipInfo"] = "\124cffFF4500Left-click:\124r Open the help page.\n\124cffFF4500Right-click:\124r Open the options menu."

L["MinimapButton"] = "Minimap button"
L["MinimapButton_Desc"] = "Shows or hides the minimap button."

L["DebugMode"] = "Debug mode"
L["DebugMode_Desc"] = "Activates or deactivates the debug mode. For normal users this is not necessary or recommended."
L["DebugMode_Activated"] = "<Debug Mode> activated."
L["DebugMode_Deactivated"] = "<Debug Mode> deactivated."

L["Profiles"] = "Profiles"


--------------------------------------------------
-- Help
--------------------------------------------------
L["Help"] = "Help"

L["Help_Opener"] = "There should be a help text here."

  
--------------------------------------------------
-- Error messages
--------------------------------------------------


modules/util.lua
local Util = GiveMeAShot:NewModule('Util')


--------------------------------------------------
-- Utility methods
--------------------------------------------------
function Util:Print(msg)
  GiveMeAShot:Print(msg)
end

function Util:PrintColored(msg, color)
  self:Print("|cff" .. color .. msg .. "|r")
end

function Util:Debug(msg)
  local Config = GiveMeAShot:GetModule('Config', true)

  if Config:GetProfile('config.debugMode') then
    self:Print(self:PrintColored('<DEBUG> ' .. msg, self:GetColor('custom', 'alert')))
  end
end

function Util:DebugOption(option, value, old)
  if old ~= nil then
    self:Debug(tostring(option)..': '..tostring(old)..' -> '..tostring(value))
  else
    self:Debug(tostring(option)..' => '..tostring(value))
  end
end

function Util:RemoveFromArray(array, removeMe)
  local j, n = 1, #array

  for i = 1, n do
    if (array[i] == removeMe) then
      array[i] = nil
    else
      -- Move i's kept value to j's position, if it's not already there.
      if (i ~= j) then
        array[j] = array[i]
        array[i] = nil
      end
      j = j + 1 -- Increment position of where we'll place the next kept value.
    end
  end

  return t
end

function Util:ShowArray(array)
  for i = 1, #array do
    print('total:'..#array , 'i:'..i, 'v:'..array[i]);
  end
end

function Util:TableContains(table, item)
  local index = 1
  while table[index] do
    if item == table[index] then
      return true
    end
    index = index + 1
  end
  return false
end

function Util:GetPartiallyColoredString(text, coloredText, color)
	local colorString = ""
	colorString = "\124cff" .. color .. coloredText .. "\124r"
	return string.format(text, colorString)
end

function Util:GetColor(parent, child)
  --https://htmlcolorcodes.com/color-names/
  --https://wowwiki.fandom.com/wiki/API_GetItemQualityColor
  local colors = {
    red = {
      indianred = "cd5c5c",
      lightcoral = "f08080",
      salmon = "fa8072",
      darksalmon = "e9967a",
      lightsalmon = "ffa07a",
      crimson = "dc143c",
      red = "ff0000",
      firebrick = "b22222",
      darkred = "8b0000",
    },
    pink = {
      pink = "ffc0cb",
      lightpink = "ffb6c1",
      hotpink = "ff69b4",
      deeppink = "ff1493",
      mediumvioletred = "c71585",
      palevioletred = "db7093",
    },
    orange = {
      lightsalmon = "ffa07a",
      coral = "ff7f50",
      tomato = "ff6347",
      orangered = "ff4500",
      darkorange = "ff8c00",
      orange = "ffa500",
    },
    yellow = {
      gold = "ffd700",
      yellow = "ffff00",
      lightyellow = "ffffe0",
      lemonchiffon = "fffacd",
      lightgoldenrodyellow = "fafad2",
      papayawhip = "ffefd5",
      moccasin = "ffe4b5",
      peachpuff = "ffdab9",
      palegoldenrod = "eee8aa",
      khaki = "f0e68c",
      darkkhaki = "bdb76b",
    },
    purple = {
      lavender = "e6e6fa",
      thistle = "d8bfd8",
      plum = "dda0dd",
      violet = "ee82ee",
      orchid = "da70d6",
      fuchsia = "ff00ff",
      magenta = "ff00ff",
      mediumorchid = "b455d3",
      mediumpurple = "9370db",
      rebeccapurple = "663399",
      blueviolet = "8a2be2",
      darkviolet = "9400d3",
      darkorchid = "9932cc",
      darkmagenta = "8b008b",
      purple = "800080",
      indigo = "4b0082",
      slateblue = "6a5acd",
      darkslateblue = "483d8b",
      mediumslateblue = "7b68ee",
    },
    green = {
      greenyellow = "adff2f",
      chartreuse = "7fff00",
      lawngreen = "fcfc00",
      lime = "00ff00",
      limegreen = "32cd32",
      palegreen = "98fb98",
      lightgreen = "90ee90",
      mediumspringgreen = "00fa9a",
      springgreen = "00ff7f",
      mediumseagreen = "3cb371",
      seagreen = "2e8b57",
      forestgreen = "228b22",
      green = "008000",
      darkgreen = "006400",
      yellowgreen = "9acd32",
      olivedrab = "6b8e23",
      olive = "808000",
      darkolivegreen = "556b2f",
      mediumaquamarine = "66cdaa",
      darkseagreen = "8fbc8b",
      lightseagreen = "20b2aa",
      darkcyan = "008b8b",
      teal = "008080",
    },
    blue = {
      aqua = "00ffff",
      cyan = "00ffff",
      lightcyan = "e0ffff",
      paleturquoise = "afeeee",
      aquamarine = "7fffd4",
      turquoise = "40e0d0",
      mediumturquoise = "48d1cc",
      darkturquoise = "00ced1",
      cadetblue = "5f9ea0",
      steelblue = "4682b4",
      lightsteelblue = "b0c4de",
      powderblue = "b0e0e6",
      lightblue = "add8e6",
      skyblue = "87ceeb",
      lightskyblue = "87cefa",
      deepskyblue = "00bfff",
      dodgerblue = "1e90ff",
      cornflowerblue = "6495ed",
      mediumslateblue = "7b68ee",
      royalblue = "4169e1",
      blue = "0000ff",
      mediumblue = "0000cd",
      darkblue = "00008b",
      navy = "000080",
      midnightblue = "191970",
    },
    brown = {
      cornsilk = "fff8dc",
      blanchedalmond = "ffebcd",
      bisque = "ffe4c4",
      navajowhite = "ffdead",
      wheat = "f5deb3",
      burlywood = "ded887",
      tan = "d2b48c",
      rosybrown = "bc8f8f",
      sandybrown = "f4a460",
      goldenrod = "daa520",
      darkgoldenrod = "b8860b",
      peru = "cd853f",
      chocolate = "d2691e",
      saddlebrown = "8b4513",
      sienna = "a0522d",
      brown = "a52a2a",
      maroon = "800000",
    },
    white = {
      white = "ffffff",
      snow = "fffafa",
      honeydew = "f0fff0",
      mintcream = "f5fffa",
      azure = "f0ffff",
      aliceblue = "f0f8ff",
      ghostwhite = "f8f8ff",
      whitesmoke = "f5f5f5",
      seashell = "fff5ee",
      beige = "f5f5dc",
      oldlace = "fdf5e6",
      floralwhite = "fffaf0",
      ivory = "fffff0",
      antiquewhite = "faebd7",
      linen = "faf0e6",
      lavenderblush = "fff0f5",
      mistyrose = "ffe4e1",
    },
    gray = {
      gainsboro = "dcdcdc",
      lightgray = "d3d3d3",
      silver = "c0c0c0",
      darkgray = "a9a9a9",
      gray = "808080",
      dimgray = "696969",
      lightslategray = "778899",
      slategray = "708090",
      darkslategray = "2f4f4f",
      black = "000000",
    },
    quality = {
      poor = "9d9d9d",
      common = "ffffff",
      uncommon = "1eff00",
      rare = "0070dd",
      epic = "a335ee",
      legendary = "ff8000",
      artifact = "e6cc80",
      heirloom = "e6cc80",
      wow_token = "00ccff",
    },
    custom = {
      info = "8BE9FD",
      success = "50FA7B",
      warning = "ffb86c",
      alert = "ff5555"
    }
  }

  return colors[tostring(parent)][tostring(child)]
end

modules/controls.lua
local Controls = GiveMeAShot:NewModule('Controls')
local L = LibStub('AceLocale-3.0'):GetLocale("GiveMeAShot")
local Util = GiveMeAShot:GetModule('Util')


--------------------------------------------------
-- UI Widget Functions
--------------------------------------------------
function Controls:createSlider(parent, name, label, description, minVal, maxVal, valStep, onValueChanged, onShow)
  local slider = CreateFrame("Slider", name, parent, "OptionsSliderTemplate")
  local editbox = CreateFrame("EditBox", name.."EditBox", slider, "InputBoxTemplate")

  slider:SetMinMaxValues(minVal, maxVal)
  slider:SetValue(minVal)
  slider:SetValueStep(1)
  slider.text = _G[name.."Text"]
  slider.text:SetText(label)
  slider.textLow = _G[name.."Low"]
  slider.textHigh = _G[name.."High"]
  slider.textLow:SetText(floor(minVal))
  slider.textHigh:SetText(floor(maxVal))
  slider.textLow:SetTextColor(0.4,0.4,0.4)
  slider.textHigh:SetTextColor(0.4,0.4,0.4)
  slider.tooltipText = label
  slider.tooltipRequirement = description

  editbox:SetSize(50,30)
  editbox:SetNumeric(true)
  editbox:SetMultiLine(false)
  editbox:SetMaxLetters(5)
  editbox:ClearAllPoints()
  editbox:SetPoint("TOP", slider, "BOTTOM", 0, -5)
  editbox:SetNumber(slider:GetValue())
  editbox:SetCursorPosition(0);
  editbox:ClearFocus();
  editbox:SetAutoFocus(false)
  editbox.tooltipText = label
  editbox.tooltipRequirement = description

	slider:SetScript("OnValueChanged", function(self,value)
		self.editbox:SetNumber(floor(value))
		if(not self.editbox:HasFocus()) then
			self.editbox:SetCursorPosition(0);
			self.editbox:ClearFocus();
		end
        onValueChanged(self, value)
	end)

  slider:SetScript("OnShow", function(self,value)
      onShow(self, value)
  end)

	editbox:SetScript("OnTextChanged", function(self)
		local value = self:GetText()

		if tonumber(value) then
			if(floor(value) > maxVal) then
				self:SetNumber(maxVal)
			end

			if floor(self:GetParent():GetValue()) ~= floor(value) then
				self:GetParent():SetValue(floor(value))
			end
		end
	end)

	editbox:SetScript("OnEnterPressed", function(self)
		local value = self:GetText()
		if tonumber(value) then
			self:GetParent():SetValue(floor(value))
				self:ClearFocus()
		end
	end)

	slider.editbox = editbox
	return slider
end

function Controls:createCheckbox(parent, name, label, description, hideLabel, onClick)
  local check = CreateFrame("CheckButton", name, parent, "InterfaceOptionsCheckButtonTemplate")
  check.label = _G[check:GetName() .. "Text"]
  if not hideLabel then
    check.label:SetText(label)
    check:SetFrameLevel(8)
  end
  check.tooltipText = label
  check.tooltipRequirement = description

  -- events
  check:SetScript("OnClick", function(self)
    local tick = self:GetChecked()
    onClick(self, tick and true or false)
  end)

  return check
end

function Controls:createEditbox(parent, name, tooltipTitle, tooltipDescription, width, height, multiline, onTextChanged)
  local editbox = CreateFrame("EditBox", name, parent, "InputBoxTemplate")
	editbox:SetSize(width, height)
	editbox:SetMultiLine(multiline)
	editbox:SetFrameLevel(9)
	editbox:ClearFocus()
  editbox:SetAutoFocus(false)
  
  if onTextChanged ~= nil then
    editbox:SetScript("OnTextChanged", function(self)
      onTextChanged(self)
    end)
  end

	editbox:SetScript("OnEnter", function(self, motion)
    Controls:ShowTooltip(self, tooltipTitle, tooltipDescription)
	end)
	editbox:SetScript("OnLeave", function(self, motion)
    Controls:HideTooltip(self)
	end)

  return editbox
end

function Controls:createLabel(parent, name, text, inheritsFrom, maxLineWidth)  
  inheritsFrom = inheritsFrom or "GameFontNormal"
  maxLineWidth = maxLineWidth or 0
  local label = parent:CreateFontString(name, "ARTWORK", inheritsFrom)  
  
  if maxLineWidth > 0  then
    local tblLines = {}
    local tblWords = {}
    local width = 0
    local str = ""
    local strSaved = ""

    -- split all single words
    for s in string.gmatch(text, "[^ ]+") do
      table.insert(tblWords, s)
    end

    -- iterate each word
    for i = 1, #tblWords do
      if str == "" then
        str = tblWords[i] 
      else 
        str = str .. " " .. tblWords[i]
      end

      -- set text to label and get width of new FontString
      label:SetText(str)
      width = label:GetStringWidth();

      -- check width and create lines
      if width <= maxLineWidth and i < #tblWords then
        strSaved = str
      else     
        if i == #tblWords then
          strSaved = str
        end

        -- create line and clear values
        table.insert(tblLines, strSaved)
        str = ""
        strSaved = ""
      end     
    end

    -- concatenate final string
    local finalLabelText = tblLines[1]
    for j = 2, #tblLines do
      finalLabelText = finalLabelText .. "\n" .. tblLines[j]
    end
    label:SetText(finalLabelText)
  else 
    label:SetText(text)
  end

  return label
end


--------------------------------------------------
-- Interface Events & Functions
--------------------------------------------------
function Controls:ShowTooltip(self, title, description)
  if self then
    GameTooltip:SetOwner(self, "ANCHOR_TOPLEFT")
    GameTooltip:SetText(title)
    GameTooltip:AddLine(description, 1, 1, 1, true)
    GameTooltip:Show()
  end
end

function Controls:HideTooltip(self)
  GameTooltip:Hide()
end

modules/config.lua
local Config = GiveMeAShot:NewModule('Config')
local L = LibStub('AceLocale-3.0'):GetLocale("GiveMeAShot")
local Util = GiveMeAShot:GetModule('Util')
local Controls = GiveMeAShot:GetModule('Controls')

--------------------------------------------------
-- Locals
--------------------------------------------------


--------------------------------------------------
-- Defaults
--------------------------------------------------
local _defaultConfig = {
  profile = {
    minimapButton = {
      hide = true,
      lock = false,
      minimapPos = 0,
    },
    config = {
      debugMode = false
    }
  },
  char = {
    level = level,
    race = race,
    class = class,
    faction = faction
  }
}


--------------------------------------------------
-- Access methods
--------------------------------------------------
function Config:GetProfile(info)
  if type(info) == 'table' then
    return self.db.profile[info[#info]]
  else
    return self.db.profile[tostring(info)]
  end
end

function Config:SetProfile(info, value)
  if type(info) == 'table' then
    local old = self.db.profile[info[#info]]
    Util:DebugOption(info[#info], value, old)
    self.db.profile[info[#info]] = value
  else
    self.db.profile[tostring(info)] = value
  end
end

function Config:GetChar(info)
  if type(info) == 'table' then
    return self.db.char[info[#info]]
  else
    return self.db.char[tostring(info)]
  end
end

function Config:SetChar(info, value)
  if type(info) == 'table' then
    local old = self.db.char[info[#info]]
    Util:DebugOption(info[#info], value, old)
    self.db.char[info[#info]] = value
  else
    self.db.char[tostring(info)] = value
  end
end

function Config:RefreshConfig(db)
  --TODO
end


--------------------------------------------------
-- Interface Events & Functions
--------------------------------------------------
function Config:SetupOptionsUI()
  GiveMeAShot.optionsFrame = CreateFrame("Frame", "GiveMeAShot_Options", InterfaceOptionsFramePanelContainer)
  GiveMeAShot.optionsFrame.name = L["Addon_Title"]
  GiveMeAShot.optionsFrame:SetAllPoints()
  HideUIPanel(GiveMeAShot.optionsFrame)

  local title = GiveMeAShot.optionsFrame:CreateFontString(nil, "ARTWORK", "GameFontNormalLarge")
  title:SetPoint("TOPLEFT", 10, -10)
  title:SetText(L["Addon_Title"])

  -- Debug Mode
  local debugModeCheckbox = Controls:createCheckbox(
    GiveMeAShot.optionsFrame,
    "GiveMeAShot_DebugMode_Checkbox",
    L["DebugMode"],
    L["DebugMode_Desc"],
    false,
    function(self, value)
      if value then
        Util:PrintColored(L["DebugMode_Activated"], Util:GetColor('custom', 'warning'))
      else
        Util:PrintColored(L["DebugMode_Deactivated"], Util:GetColor('custom', 'warning'))
      end
      Config:SetProfile('config.debugMode', value)
    end
  )
  debugModeCheckbox:SetPoint("TOPRIGHT", GiveMeAShot.optionsFrame, -100, -10)

  -- Minimap Button
  local minimapButtonCheckbox = Controls:createCheckbox(
    GiveMeAShot.optionsFrame,
    "GiveMeAShot_MinimapButton_Checkbox",
    L["MinimapButton"],
    L["MinimapButton_Desc"],
    false,
    function(self, value)
      GiveMeAShot:ToggleMinimapButton()
    end
  )
  minimapButtonCheckbox:SetChecked(not Config:GetProfile('minimapButton.hide'))
  minimapButtonCheckbox:SetPoint("TOPLEFT", title, 10, -30)

  -- add to interface options
  InterfaceOptions_AddCategory(GiveMeAShot.optionsFrame);
end

function Config:SetupHelpUI()
  local MAX_FRAME_WIDTH = 550

  GiveMeAShot.helpFrame = CreateFrame("Frame", "GiveMeAShot_Help", GiveMeAShot.optionsFrame)
  GiveMeAShot.helpFrame.name = L["Help"]
  GiveMeAShot.helpFrame.parent = GiveMeAShot.optionsFrame.name
  GiveMeAShot.helpFrame:SetAllPoints()
  HideUIPanel(GiveMeAShot.helpFrame)

  local title = GiveMeAShot.helpFrame:CreateFontString(nil, "ARTWORK", "GameFontNormalLarge")
  title:SetPoint("TOPLEFT", 10, -10)
  title:SetText(L["Addon_Title"] .. " :: " .. L["Help"])

  -- Opener
  local helpLabelOpener = Controls:createLabel(GiveMeAShot.helpFrame, "GiveMeAShot_HelpLabel_Opener", L["Help_Opener"], "GameFontNormal", MAX_FRAME_WIDTH)
  helpLabelOpener:SetPoint("TOPLEFT", title, "BOTTOMLEFT", 0, -20)

  -- add to interface options
  InterfaceOptions_AddCategory(GiveMeAShot.helpFrame);
end


--------------------------------------------------
-- INIT
--------------------------------------------------
function Options:OnInitialize()
  -- register database
  self.db = LibStub('AceDB-3.0'):New('GiveMeAShotDB', _defaultConfig)
  self.db:RegisterDefaults(_defaultConfig)

  -- setup profile options
  profileOptions = LibStub("AceDBOptions-3.0"):GetOptionsTable(self.db)
  LibStub("AceConfig-3.0"):RegisterOptionsTable("GiveMeAShot_Profiles", profileOptions)
  profileSubMenu = LibStub("AceConfigDialog-3.0"):AddToBlizOptions("GiveMeAShot_Profiles", L["Profiles"], L["Addon_Title_Short"])
  
  self.db.RegisterCallback(self, "OnProfileChanged", "RefreshConfig")
  self.db.RegisterCallback(self, "OnProfileCopied", "RefreshConfig")
  self.db.RegisterCallback(self, "OnProfileReset", "RefreshConfig")
end